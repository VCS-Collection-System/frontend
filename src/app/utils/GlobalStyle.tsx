import * as React from 'react';
import { createGlobalStyle } from 'styled-components';

const GlobalStyle = () => {

  const fontColor = "rgb(180, 180, 178)"; 
  const backgroundColor200 = "var(--pf-global--BackgroundColor--dark-200)";
  const backgroundColor300 = "var(--pf-global--BackgroundColor--dark-300)";
  const backgroundColor400 = "var(--pf-global--BackgroundColor--dark-400)"; 
  const borderColor200 = "var(--pf-global--BorderColor--200)"; 

  const GlobalStyles = createGlobalStyle`
    .pf-c-page__main {
      background-color: ${backgroundColor400}
    }
    .pf-c-button.pf-m-plain#span {
      color: ${fontColor};
    }
    .pf-c-title.pf-m-3xl { 
      color: ${fontColor};
    }
    .pf-c-card {
      --pf-c-card--BackgroundColor: ${backgroundColor200};
      color: ${fontColor};
    }
    .pf-c-card .pf-c-card{
      --pf-c-card--BackgroundColor: ${backgroundColor400};
      color: ${fontColor};
    }
    .pf-c-card .pf-c-card .pf-c-card{
      --pf-c-card--BackgroundColor: ${backgroundColor200};
      color: ${fontColor};
    }
    .pf-c-form-control:disabled {
      --pf-c-form-control--BackgroundColor: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-form-control {
      --pf-c-form-control--BackgroundColor: ${backgroundColor300};
      --pf-c-form-control--BorderTopColor: ${backgroundColor300};
      --pf-c-form-control--BorderRightColor: ${backgroundColor300};
      --pf-c-form-control--BorderBottomColor: ${backgroundColor300};
      --pf-c-form-control--BorderLeftColor: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-button.pf-m-tertiary {
      color: ${fontColor};
    };
    .pf-c-button.pf-m-control {
      background-color: ${backgroundColor300};
      --pf-c-button--after--BorderColor: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-calendar-month {
      --pf-c-calendar-month--BackgroundColor: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-calendar-month__dates-cell::before {
      background-color: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-calendar-month__date {
      color: ${fontColor};
    };
    .pf-c-select__toggle {
      background-color: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-form-control[readonly] {
      background-color: ${backgroundColor300};
      color: ${fontColor};
    };
    .pf-c-form__helper-text {
      color: ${fontColor};
    };
    .pf-c-menu {
      --pf-c-menu--BackgroundColor: ${backgroundColor300};
      color: color: ${fontColor};
    };
    .pf-c-menu__item {
      background-color: ${backgroundColor300};
      color: ${fontColor};
    }
    .pf-c-input-group {
      --pf-c-input-group--BackgroundColor: ${backgroundColor300};
    }
    .pf-c-search-input {
      --pf-c-search-input__text--before--BorderColor: ${borderColor200};
      --pf-c-search-input__text--after--BorderBottomColor: ${borderColor200};
      color: ${fontColor};
    }
    .pf-c-search-input__text-input {
      color: ${fontColor};
    }
    .pf-c-modal-box {
      --pf-c-modal-box--BackgroundColor: ${backgroundColor300};
      color: ${fontColor};
    }
    .pf-c-radio {
      --pf-c-radio__label--Color: ${fontColor};
    }
    .pf-c-button.pf-m-plain {
      color: ${fontColor};
    }
    `;

    return <GlobalStyles />
};

export default GlobalStyle;