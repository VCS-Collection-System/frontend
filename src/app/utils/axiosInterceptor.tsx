import axios from "axios";

const axiosInterceptor = axios.create({})

axiosInterceptor.interceptors.request.use(
    config => {
        config.headers['Authorization'] = 'Bearer ' + sessionStorage.getItem("token");
        return config;
    },
    error => {
        Promise.reject(error)
    });
export default axiosInterceptor;