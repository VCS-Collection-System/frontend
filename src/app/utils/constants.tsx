declare global {  
  interface Window {
    REACT_APP_PAM_URL_BASE: string;
    REACT_APP_PAM_CONTAINER_ID: string;
    REACT_APP_KEYCLOAK_COUNTRY: string;
    REACT_APP_KEYCLOAK_EMAIL: string;
    REACT_APP_KEYCLOAK_FIRST_NAME: string;
    REACT_APP_KEYCLOAK_ID: string;
    REACT_APP_KEYCLOAK_LAST_NAME: string;
    REACT_APP_MAINTENANCE_BANNER: boolean;
    REACT_APP_MAINTENANCE_BANNER_TEXT: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    REACT_APP_FEATURE_FLAGS: Array<Record<string, any>>;
    REACT_APP_MAX_DOSES: Record<string, number>;
    REACT_APP_CONSENT_LABEL: string;
    REACT_APP_ALTERNATE_CONSENT_LABEL: string;
    REACT_APP_BLOCKED_COUNTRIES: Array<string>;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    REACT_APP_GLOBAL_CONFIG: Record<string, any>;
  }
}

// PAM urls
export const PAM_BASE_URL = window.REACT_APP_PAM_URL_BASE; 
export const SUBMIT_VAX_URL = PAM_BASE_URL + '/attestation/vax';
export const SUBMIT_TEST_URL = PAM_BASE_URL + '/attestation/covid-test-result';
export const PAM_TASK_URL = PAM_BASE_URL + '/tasks/';
export const RESERVED_TASK_URL = PAM_TASK_URL + 'owners?status=Reserved&user=';
export const INPROGRESS_TASK_URL = PAM_TASK_URL + 'owners?status=InProgress&user=';
export const ALL_TASK_URL = PAM_TASK_URL + 'owners?status=Ready&pageSize=200';
export const PRIORITY_URL = PAM_BASE_URL + '/query/priority-results';
export const PRIORITY_CONFIRM_URL = PAM_BASE_URL + '/query/priority-result/';
export const ATTACHMENT_URL = PAM_BASE_URL + '/attestation/attachment';
export const EMPLOYEE_URL = PAM_BASE_URL + '/query/employee/';
export const EMPLOYEE_NAME_URL = PAM_BASE_URL + '/query/employee-name/';
export const EMPLOYEE_VAX_HISTORY_URL = PAM_BASE_URL + '/query/vax-document/';
export const EMPLOYEE_TEST_HISTORY_URL = PAM_BASE_URL + '/query/test-result-document/';
export const VAX_TASK_DOCUMENT_URL = PAM_BASE_URL + '/query/task/document/';
export const DELETE_VAX_URL = PAM_BASE_URL + '/attestation/delete-vax-record/';
export const CONFIG_URL = PAM_BASE_URL + '/query/country-configs/';
export const SAVE_CONFIG_URL = PAM_BASE_URL + '/attestation/country-configs';

// List of countries allow to use this app
export const BLOCKED_COUNTRIES = window.REACT_APP_BLOCKED_COUNTRIES || []

export const KONAMI_CODE = [ "ArrowUp", "ArrowUp", "ArrowDown", "ArrowDown", "ArrowLeft", "ArrowRight", "ArrowLeft", "ArrowRight", "KeyB", "KeyA"];

// Maintenance Banner
export const MAINTENANCE_BANNER = window.REACT_APP_MAINTENANCE_BANNER || false
export const MAINTENANCE_BANNER_TEXT = window.REACT_APP_MAINTENANCE_BANNER_TEXT || ""

// Keycloak claims
export const KEYCLOAK_COUNTRY = window.REACT_APP_KEYCLOAK_COUNTRY
export const KEYCLOAK_EMAIL = window.REACT_APP_KEYCLOAK_EMAIL
export const KEYCLOAK_FIRST_NAME = window.REACT_APP_KEYCLOAK_FIRST_NAME
export const KEYCLOAK_ID = window.REACT_APP_KEYCLOAK_ID
export const KEYCLOAK_LAST_NAME = window.REACT_APP_KEYCLOAK_LAST_NAME
export const FEATURE_FLAGS = window.REACT_APP_FEATURE_FLAGS || []

// Maximum doses allowed by vaccine manufacturer
export const MAX_DOSES = window.REACT_APP_MAX_DOSES || {}

export const CONSENT_LABEL = window.REACT_APP_CONSENT_LABEL
export const ALTERNATE_CONSENT_LABEL = window.REACT_APP_ALTERNATE_CONSENT_LABEL

// Global country config
export const GLOBAL_CONFIG = window.REACT_APP_GLOBAL_CONFIG || {};

