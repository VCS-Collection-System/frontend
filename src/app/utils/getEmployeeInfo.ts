import {
  KEYCLOAK_FIRST_NAME, 
  KEYCLOAK_LAST_NAME, 
  KEYCLOAK_EMAIL,
  KEYCLOAK_COUNTRY,
  KEYCLOAK_ID
} from "@app/utils/constants"; 

export function getEmployeeInfo(keycloak) {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const jp = require('jsonpath');
  const profile = keycloak.profile;
  const employeeFirstName = jp.query(profile, KEYCLOAK_FIRST_NAME)[0];
  const employeeLastName = jp.query(profile, KEYCLOAK_LAST_NAME)[0];
  const employeeEmail = jp.query(profile, KEYCLOAK_EMAIL)[0];
  const employeeCountry = jp.query(profile, KEYCLOAK_COUNTRY)[0];
  const employeeID = jp.query(profile, KEYCLOAK_ID)[0];

  const isApprover = () => {
    if(employeeCountry && 
      (keycloak.hasRealmRole("vcs-superapprover") || keycloak.hasRealmRole("vcs-approver-" + employeeCountry.toLowerCase()))){
        return true;
    }

    return false;
  }

  const isSuperApprover = () => {
    return keycloak.hasRealmRole("vcs-superapprover");
  }

  return {
    firstName: employeeFirstName, 
    lastName: employeeLastName,
    email: employeeEmail,
    country: employeeCountry,
    id: employeeID,
    isApprover: isApprover(),
    isSuperApprover: isSuperApprover()
  }
}
