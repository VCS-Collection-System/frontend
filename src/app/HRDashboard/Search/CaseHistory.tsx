import React, { useEffect, useState } from 'react';
import { 
  Accordion,
  AccordionContent,
  AccordionItem,
  AccordionToggle,
  Button,
  Card, 
  CardBody, 
  CardFooter,
  Grid,
  Flex,
  FlexItem,
  GridItem,
  Modal, 
  ModalVariant,
  Menu, 
  MenuContent, 
  MenuList, 
  MenuItem,
  MenuGroup,
  Split
} from '@patternfly/react-core';
import CardWithTitle from '@app/Components/CardWithTitle';
import { useHistory } from 'react-router-dom';
import authaxios from "@app/utils/axiosInterceptor";
import { HRVaxSubmit } from './AddRecord/HRVaxSubmit';
import { HRTestSubmit } from './AddRecord/HRTestSubmit';
import { dateConvert } from "@app/utils/utils";
import EmployeeInfoHorizontalCard from '@app/Components/EmployeeInfoHorizontalCard';
import { CaseReviewPanel } from './CaseReviewPanel';
import { EMPLOYEE_VAX_HISTORY_URL, EMPLOYEE_TEST_HISTORY_URL } from '@app/utils/constants'

const CaseHistory: React.FunctionComponent = () => {
  const history = useHistory();
  const [isModalOpen, setIsModalOpen] = useState(false); 
  const [employeeId, setEmployeeId] = useState("");
  const [employeeName, setEmployeeName] = useState("");
  const [employeeDOB, setEmployeeDOB] = useState("");
  const [employeeEmail, setEmployeeEmail] = useState("");
  const [refresh, setRefresh] = useState(false);
  const [activeItem, setActiveItem] = useState('0'); 
  const [vaxHistory, setVaxHistory] = useState([]);
  const [testHistory, setTestHistory] = useState([]);  
  const [hasSelectedItem, setHasSelectedItem] = useState(false); 
  const [updatePanel, setUpdatePanel] = useState(false);
  const [expanded, setExpanded] = useState('');

  useEffect(() => {
    const sessionEmployee = sessionStorage.getItem("searchResultInfo") || "{}"; 
    const employeeInfo = JSON.parse(sessionEmployee); 
    setEmployeeId(employeeInfo['employeeId']);
    setEmployeeName(employeeInfo['firstName'] + " " + employeeInfo['lastName']);
    setEmployeeDOB(employeeInfo['dateOfBirth']);
    setEmployeeEmail(employeeInfo['email']);

    authaxios.get(EMPLOYEE_VAX_HISTORY_URL + employeeInfo['employeeId'])
    .then(result => {
      result['data'].sort((a, b) => {
        return a['vaccineAdministrationDate'].localeCompare(b['vaccineAdministrationDate']);
      });
      setVaxHistory(result['data'])
    })

    authaxios.get(EMPLOYEE_TEST_HISTORY_URL + employeeInfo['employeeId'])
    .then(result => {
      setTestHistory(result['data'])
    })

  }, [refresh]);

  const closeModal = () => {
    setIsModalOpen(false);
    setRefresh(!refresh);
    setUpdatePanel(!updatePanel);
  }

  const onSelect = (event, itemId) => {
    setHasSelectedItem(true);
    setActiveItem(itemId);
    setUpdatePanel(!updatePanel)

    if(itemId.includes('TEST')){
      const dataIndex = parseInt(itemId.split(':')[1]);
      const testData = testHistory[dataIndex];
      sessionStorage.setItem("caseHistoryType", "TEST");
      sessionStorage.setItem("caseHistoryInfo", JSON.stringify(testData));
    }
    else if(itemId.includes('VAX')) {
      const dataIndex = parseInt(itemId.split(':')[1]);
      const vaxData = vaxHistory[dataIndex];
      sessionStorage.setItem("caseHistoryType", "VAX");
      sessionStorage.setItem("caseHistoryInfo", JSON.stringify(vaxData));
    }
  }

  const getTaskStatus = (value) => {
    if(value['review']) {
      return value['review']['reviewOutcome'];
    }
    return value['autoApproved'] ? "AUTO-ACCEPTED" : "NOT REVIEWED";
  }

  const AddRecordModal = () => {
    function onToggle(id) {
      if(id === expanded) {
        setExpanded('');
      }
      else {
        setExpanded(id); 
      }
    }


    return (
       <Modal
        aria-label="AddRecordModal"
        variant={ModalVariant.medium}
        title="Add Record"
        isOpen={isModalOpen}
        onClose={closeModal}
      >
        <Accordion>
          <AccordionItem>
            <AccordionToggle
              onClick={() => {
                onToggle('vax-toggle');
              }}
              isExpanded={expanded === 'vax-toggle'}
              id="vax-toggle"
            >
              Vaccine Submission
            </AccordionToggle>
            <AccordionContent id="vax-expand" isHidden={expanded !== 'vax-toggle'}>
              <HRVaxSubmit closeParentModal={closeModal}/>
            </AccordionContent>
          </AccordionItem>
          <AccordionItem>
            <AccordionToggle
              onClick={() => {
                onToggle('test-toggle');
              }}
              isExpanded={expanded === 'test-toggle'}
              id="test-toggle"
            >
              Test Submission
            </AccordionToggle>
            <AccordionContent id="test-expand" isHidden={expanded !== 'test-toggle'}>
              <HRTestSubmit closeParentModal={closeModal}/>
            </AccordionContent>
          </AccordionItem>
        </Accordion>
        <br />
        <Flex>
          <FlexItem align={{ default: "alignLeft" }}>
            <Button key="back" variant="plain" onClick={closeModal}>
              Back
            </Button>
          </FlexItem>
        </Flex>
      </Modal>
    )
  }

  const CovidTestHistory = () => {
    if(testHistory.length > 0) {
      return (
        <>
          <MenuItem 
            key={"TEST"} 
            itemId={"TEST"}
            isDisabled
          >
            <Flex justifyContent={{default: "justifyContentSpaceBetween"}}>
              <FlexItem>
              {"Test Date"}
              </FlexItem>
              <FlexItem>
              {"Test Result"}
              </FlexItem>
            </Flex>
          </MenuItem>
          {
            testHistory.map((x, i) => (
              <MenuItem 
                key={"TEST:" + i} 
                itemId={"TEST:"+i}
              >
                <Flex justifyContent={{default: "justifyContentSpaceBetween"}}>
                  <FlexItem>
                    {dateConvert(x['covidTestDate'])}
                  </FlexItem>
                  <FlexItem>
                  { x['covidTestResult'] }
                    {/* { x['covidTestResult'] === "POSITIVE" ? <CircleIcon color='firebrick'/>  : x['covidTestResult'] === "NEGATIVE" ?  <CircleIcon color="seagreen" />: <CircleIcon color="goldenrod"/> } */}
                  </FlexItem>
                </Flex>
              </MenuItem>
            ))
          }
        </>
      )
    }
    else{
      return (
        <MenuItem key="empty" itemId="empty" isDisabled>
          [No Test Results]
        </MenuItem>
      )
    }
  }

  const CovidVaccineHistory = () => {
    if(vaxHistory.length > 0) {
      return (
        <>
          <MenuItem 
            key={"VAX"} 
            itemId={"VAX"}
            isDisabled
          >
            <Flex justifyContent={{default: "justifyContentSpaceBetween"}}>
              <FlexItem>
                {"Vaccination Date"}
              </FlexItem>
              <FlexItem>
                {"Status"} 
              </FlexItem>
            </Flex>
          </MenuItem>
          {
            vaxHistory.map((value, index) => (
              <MenuItem 
                key={"VAX:" + index} 
                itemId={"VAX:" + index}
              >
                <Flex justifyContent={{default: "justifyContentSpaceBetween"}}>
                  <FlexItem>
                    { dateConvert(value['vaccineAdministrationDate']) }
                  </FlexItem>
                  <FlexItem>
                    { getTaskStatus(value) }
                  </FlexItem>
                </Flex>
              </MenuItem>
            ))
          }
        </>
      )
    }
    else{
      return (
        <MenuItem key="empty" itemId="empty" isDisabled>
          [No Vaccination Records]
        </MenuItem>
      )
    }
  }

  return (
    <Card isRounded>
      <AddRecordModal />
      <CardWithTitle title="Case History" info="Review Case History"/>
      <CardBody>
        <Grid hasGutter>
          <GridItem span={12}> 
            <EmployeeInfoHorizontalCard 
              employeeId={employeeId}
              employeeName={employeeName}
              employeeDOB={employeeDOB}
              employeeEmail={employeeEmail}
            />
          </GridItem>
          <GridItem span={4} rowSpan={6}> 
            <Card>
              <CardWithTitle title="Attestation Records"/>
                <Menu onSelect={onSelect} activeItemId={activeItem}>
                  <MenuContent menuHeight="875px">
                    <MenuGroup label="Test Results">
                      <MenuList className="indentList">
                        <CovidTestHistory />
                      </MenuList>
                    </MenuGroup>
                    <MenuGroup label="Vaccination Records">
                      <MenuList className="indentList">
                        <CovidVaccineHistory />
                      </MenuList>
                    </MenuGroup>
                  </MenuContent>
                </Menu>
            </Card>
          </GridItem>
          <GridItem span={8} rowSpan={6}>
            <CaseReviewPanel 
              hasSelectedItem={hasSelectedItem}
              setHasSelectedItem={setHasSelectedItem}
              updatePanel={updatePanel} 
              refreshCaseHistory={refresh}
              setResfreshCaseHistory={setRefresh}
            />
          </GridItem>
        </Grid>
      </CardBody>
      <CardFooter>
        <Split hasGutter>
          <Button variant="primary" onClick = { () => setIsModalOpen(!isModalOpen) }>
            Add Record
          </Button>
          <Button 
            variant="plain" 
            onClick={() => history.push("/hrdashboard")
          }>
            Back
          </Button>
        </Split>
      </CardFooter>
    </Card>
  );
}

export { CaseHistory };