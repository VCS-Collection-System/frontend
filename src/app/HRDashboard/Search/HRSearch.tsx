import React, { useState } from 'react';
import { 
  Menu, 
  MenuContent, 
  MenuList, 
  MenuItem, 
  SearchInput, 
  Card, 
  Button, 
  GridItem, 
  Grid
} from '@patternfly/react-core';
import { useHistory } from 'react-router-dom';
import CardWithTitle from '@app/Components/CardWithTitle';
import authaxios from "@app/utils/axiosInterceptor";
import { EMPLOYEE_URL } from '@app/utils/constants'

const HRSearch: React.FunctionComponent = () => {
  const [activeItem, setActiveItem] = useState('0'); 
  const [search, setSearch] = useState(''); 
  const [searchResult, setSearchResult] = useState<string[]>([])
  const [searchResultInfo, setSearchResultInfo] = useState<string[]>([])
  const [hasNoResult, setHasNoResult] = useState(false)
  const history = useHistory();
  
  const onSelect = (event, itemId) => {
    const itemIndex = parseInt(itemId)
    sessionStorage.setItem("searchResultInfo", JSON.stringify(searchResultInfo[itemIndex]))
    setActiveItem(itemId)
    history.push('/hrdashboard/history');
  }

  const onSearchChange = (value) => {
    setSearch(value);
    setSearchResult([])
    setHasNoResult(false)
  }

  const onSearchClear = () => {
    setSearch('')
    setSearchResult([])
    setHasNoResult(false)
  }

  const onSearchEnter = () => {
    authaxios.get(EMPLOYEE_URL + search)
    .then(result => {
      if(result['data']) {
        const employee = result['data'];
        const resultLabel = employee['lastName'] + ", " + employee['firstName'] + "  [" + employee['employeeId'] + "]";
        
        setHasNoResult(false)
        setSearchResult([resultLabel])
        setSearchResultInfo([employee])
      }
      else {
        setHasNoResult(true)
      }
    })
    .catch(() => {
      setHasNoResult(true)
    })
  }

  return (
    <Card isRounded>
      <CardWithTitle title="Red Hat VCS Search" info="Search Employee" />
      <Menu onSelect={onSelect} activeItemId={activeItem}>
        <Grid>
          <GridItem span={9}>
            <SearchInput
              placeholder='Search by Employee ID'
              value={search}
              onChange={onSearchChange}
              onClear={onSearchClear}
              onKeyPress={(event) => {
                if(event.key === 'Enter') {
                  onSearchEnter()
                }
              }}
            />
          </GridItem>
          <GridItem span={3}>
            <Button isBlock variant="control" onClick={onSearchEnter}>Search</Button>
          </GridItem>
        </Grid> 
        <MenuContent menuHeight="315px">
          <MenuList>
            <MenuItem isDisabled itemId="title">
              Search Result:
            </MenuItem>
            {
              searchResult &&
              searchResult.map((data, index) =>
              <MenuItem key={index} itemId={index}>
                {data}
              </MenuItem> ) 
            }
            {
              hasNoResult &&
              <MenuItem isDisabled itemId='NoResults'>
                No results for {search}
              </MenuItem>
            }
          </MenuList>
        </MenuContent>
      </Menu>
    </Card>
  );
};

export { HRSearch };