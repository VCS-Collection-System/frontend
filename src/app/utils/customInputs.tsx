import { useEffect, useState } from "react";
import { KONAMI_CODE } from './constants'

const useInputEvent = () => {
  const [key, setKey] = useState(null);
  useEffect(() => {
    const keyDownHandler = ({ code }) => setKey(code);
    const keyUpHandler = () => setKey(null);
    window.addEventListener('keydown', keyDownHandler);
    window.addEventListener('keyup', keyUpHandler);
    return () => {
      window.removeEventListener("keydown", keyDownHandler);
      window.removeEventListener("keyup", keyUpHandler)
    }
  }, []);
  return key;
}

const useSecretCode = (secretCode) => {
  const [count, setCount] = useState(0);
  const [success, setSuccess] = useState(false);
  const key = useInputEvent();
  
  useEffect(() => {
    if (key == null) return;
    if (key !== secretCode[count]) {
      setCount(0);
      return;
    }
    
    setCount((state) => state + 1);
    if (count + 1 === secretCode.length) {
      setSuccess(!success);
      setCount(0);
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [key]); 
  
  return success;
};

export const useDarkMode = () => {
  return useSecretCode(KONAMI_CODE);
};
