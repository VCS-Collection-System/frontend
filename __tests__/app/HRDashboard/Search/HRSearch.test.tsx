import React from 'react';
import { HRSearch } from '../../../../src/app/HRDashboard/Search/HRSearch';
import { act } from 'react-dom/test-utils';
import { EMPLOYEE_URL, EMPLOYEE_NAME_URL } from '../../../../src/app/utils/constants'
import { fireEvent, render, screen } from '@testing-library/react'
import reactRouterDom from 'react-router-dom'

jest.mock('react-router-dom')
const mockPush = jest.fn()
reactRouterDom.useHistory = jest.fn().mockReturnValue({push: mockPush})

jest.mock("@app/utils/constants", () => ({
  EMPLOYEE_URL: "employee-id",
  EMPLOYEE_NAME_URL: "employee-name"
}))

jest.mock('@app/utils/axiosInterceptor', () => ({
  get: (URL) => {
    let returnData;
    if(URL === EMPLOYEE_NAME_URL + 'validName'){
      returnData = {data: [{lastName: "mockLast", firstName: "mockFirst", employeeId: "validID"}]}
    }
    else if(URL === EMPLOYEE_NAME_URL + 'invalidName'){
      return Promise.reject({data: {}})
    }
    else if(URL === EMPLOYEE_URL + 'validID'){
      returnData = {data: {lastName: "mockLast", firstName: "mockFirst", employeeId: "validID"}}
    }
    else if(URL === EMPLOYEE_URL + 'invalidID'){
      return Promise.reject({data: {}})
    }
    else{
      returnData = {};
    }

    return Promise.resolve(returnData)
  },
}))

afterEach(() => {
  sessionStorage.clear(); 
})

describe('<HRSearch />', ()=> {
  it('should be defined', () => {
    expect(HRSearch).toBeDefined()
  })

  it("should match existing snapshot", async () => {
    const { asFragment } = render(<HRSearch />)
    expect(asFragment()).toMatchSnapshot()
  })

  it("should return with No results for invalid user", async () => {
    render(<HRSearch />)

    const searchInput = screen.getByRole('textbox', {name: /Search input/i})
    const searchButton = screen.getAllByRole('button', {name: /Search/i})[0]
    fireEvent.change(searchInput, {target: {value: 'mockInput'}})
    await act(async () => {
      fireEvent.click(searchButton)
    })

    expect(screen.getByText('No results for mockInput')).toBeTruthy(); 
  })

  it("should return results for valid user and clickable", async () => {
    render(<HRSearch />)

    const searchInput = screen.getByRole('textbox', {name: /Search input/i})
    const searchButton = screen.getAllByRole('button', {name: /Search/i})[0]
    fireEvent.change(searchInput, {target: {value: 'validID'}})
    await act(async () => {
      fireEvent.click(searchButton)
    })

    expect(screen.getByText('mockLast, mockFirst [validID]')).toBeTruthy(); 

    const resultButton = screen.getByRole('button', {name: "mockLast, mockFirst [validID]"})
    await act(async () => {
      fireEvent.click(resultButton)
    })

    expect(mockPush).toHaveBeenCalledWith('/hrdashboard/history')
  })

  it("should return results for valid user when hitting enter", async () => {
    render(<HRSearch />)

    const searchInput = screen.getByRole('textbox', {name: /Search input/i})
    fireEvent.change(searchInput, {target: {value: 'validID'}})
    await act(async () => {
      fireEvent.keyPress(searchInput, { key: "Enter", keyCode: 13, code: 13, charCode: 13 })
    })

    expect(screen.getByText('mockLast, mockFirst [validID]')).toBeTruthy(); 
  })

  // it("should return results for valid user when searching with last name", async () => {
  //   render(<HRSearch />)

  //   const searchInput = screen.getByRole('textbox', {name: /Search input/i})
  //   fireEvent.change(searchInput, {target: {value: 'validName'}})
  //   await act(async () => {
  //     fireEvent.keyPress(searchInput, { key: "Enter", keyCode: 13, code: 13, charCode: 13 })
  //   })

  //   expect(screen.getByText('mockLast, mockFirst [validID]')).toBeTruthy(); 
  // })

  it("should not return results for valid user when not hitting enter", async () => {
    render(<HRSearch />)

    const searchInput = screen.getByRole('textbox', {name: /Search input/i})
    fireEvent.change(searchInput, {target: {value: 'validID'}})
    await act(async () => {
      fireEvent.keyPress(searchInput, { key: "a", keyCode: 65, code: 65, charCode: 65})
    })
    expect(screen.queryAllByText('mockLast, mockFirst [validID]')).toEqual([]);
    expect(mockPush).toBeCalledTimes(0)
  })

  it("should return no result if API errors", async () => {
    render(<HRSearch />)

    const searchInput = screen.getByRole('textbox', {name: /Search input/i})
    const searchButton = screen.getAllByRole('button', {name: /Search/i})[0]
    fireEvent.change(searchInput, {target: {value: 'invalidID'}})
    await act(async () => {
      fireEvent.click(searchButton)
    })

    expect(screen.getByText('No results for invalidID')).toBeTruthy(); 
  })

  // it("should return no result if Last name API errors", async () => {
  //   render(<HRSearch />)

  //   const searchInput = screen.getByRole('textbox', {name: /Search input/i})
  //   const searchButton = screen.getAllByRole('button', {name: /Search/i})[0]
  //   fireEvent.change(searchInput, {target: {value: 'invalidName'}})
  //   await act(async () => {
  //     fireEvent.click(searchButton)
  //   })

  //   expect(screen.getByText('No results for invalidName')).toBeTruthy(); 
  // })

  it("should be able to clear search input", async () => {
    render(<HRSearch />)

    const searchInput = screen.getByRole('textbox', {name: /Search input/i})
    await act(async () => {
      fireEvent.change(searchInput, {target: {value: 'invalidID'}})
    })
    const resetButton = screen.getByRole('button', {name: /Reset/i})
    expect(resetButton).toBeTruthy(); 
    await act(async () => {
      fireEvent.click(resetButton)
    })
    expect(searchInput.getAttribute('value')).toEqual("")
  })
})