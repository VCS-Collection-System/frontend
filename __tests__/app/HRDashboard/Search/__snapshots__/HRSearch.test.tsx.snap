// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HRSearch /> should match existing snapshot 1`] = `
<DocumentFragment>
  <article
    class="pf-c-card pf-m-rounded"
    data-ouia-component-id="OUIA-Generated-Card-1"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__title"
    >
      <p
        class=""
        data-pf-content="true"
      >
        Red Hat VCS Search
      </p>
      <div
        class="pf-c-content"
      >
        <small
          class=""
          data-pf-content="true"
        >
          Search Employee
        </small>
      </div>
    </div>
    <div
      aria-label="Global"
      class="pf-c-menu"
      data-ouia-component-id="OUIA-Generated-Menu-1"
      data-ouia-component-type="PF4/Menu"
      data-ouia-safe="true"
    >
      <div
        class="pf-l-grid"
      >
        <div
          class="pf-l-grid__item pf-m-9-col"
        >
          <div
            class="pf-c-search-input"
          >
            <div
              class="pf-c-input-group"
            >
              <div
                class="pf-c-search-input__bar"
              >
                <span
                  class="pf-c-search-input__text"
                >
                  <span
                    class="pf-c-search-input__icon"
                  >
                    <svg
                      aria-hidden="true"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      style="vertical-align: -0.125em;"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                      />
                    </svg>
                  </span>
                  <input
                    aria-label="Search input"
                    class="pf-c-search-input__text-input"
                    placeholder="Search by Employee ID"
                    value=""
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="pf-l-grid__item pf-m-3-col"
        >
          <button
            aria-disabled="false"
            class="pf-c-button pf-m-control pf-m-block"
            data-ouia-component-id="OUIA-Generated-Button-control-1"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            tabindex="0"
            type="button"
          >
            Search
          </button>
        </div>
      </div>
      <div
        class="pf-c-menu__content"
        style="--pf-c-menu__content--Height: 315px;"
      >
        <ul
          class="pf-c-menu__list"
        >
          <li
            class="pf-c-menu__list-item pf-m-disabled"
            tabindex="-1"
          >
            <button
              aria-current="false"
              class="pf-c-menu__item"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                class="pf-c-menu__item-main"
              >
                <span
                  class="pf-c-menu__item-text"
                >
                  Search Result:
                </span>
              </span>
            </button>
          </li>
        </ul>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
